micrometer-tracing/src/main/java/io/micrometer/tracing/handler/RevertingScope.java

```diff
@@ -64,4 +70,53 @@ public int hashCode() {
         return Objects.hash(tracingContext, currentScope, previousScope);
     }
 
+    static RevertingScope maybeWithBaggage(Tracer tracer, TracingContext tracingContext,
+            @Nullable TraceContext newContext, RevertingScope revertingScopeForSpan,
+            Scope previousScopeOnThisObservation) {
+        RevertingScope revertingScope = revertingScopeForSpan;
+        ContextView context = tracingContext.getContext();
+        if (context == null) {
+            return revertingScope;
+        }
+        Set<KeyValue> baggageKeyValues = matchingBaggageKeyValues(tracer, context);
+        if (baggageKeyValues.isEmpty()) {
+            return revertingScope;
+        }
+        ArrayDeque<BaggageInScope> scopes = startBaggageScopes(tracer, newContext, baggageKeyValues);
+        return new RevertingScope(tracingContext, () -> {
+            for (BaggageInScope scope : scopes) {
+                scope.close();
+            }
+            revertingScope.close();
+        }, previousScopeOnThisObservation);
+    }
+
+    private static ArrayDeque<BaggageInScope> startBaggageScopes(Tracer tracer, TraceContext newContext,
+            Set<KeyValue> baggageKeyValues) {
+        ArrayDeque<BaggageInScope> scopes = new ArrayDeque<>();
+        for (KeyValue keyValue : baggageKeyValues) {
+            if (newContext != null) {
+                scopes.addFirst(tracer.createBaggageInScope(newContext, keyValue.getKey(), keyValue.getValue()));
+            }
+            else {
+                scopes.addFirst(tracer.createBaggageInScope(keyValue.getKey(), keyValue.getValue()));
+            }
+        }
+        return scopes;
+    }
+
+    private static Set<KeyValue> matchingBaggageKeyValues(Tracer tracer, ContextView context) {
+        Set<String> lowerCaseRemoteFields = new HashSet<>();
+        for (String remoteField : tracer.getBaggageFields()) {
+            lowerCaseRemoteFields.add(remoteField.toLowerCase());
+        }
+        Set<KeyValue> baggageKeyValues = new HashSet<>();
```

Comment:
nit: Different from the `lowerCaseRemoteFields` variable, this `baggageKeyValues` variable (and the return type of this method, etc) might be better to keep as `List`. That way, creating `BaggageInScope` is always in a predictable order which might give an easier debug experience, just in case. If the ordering doesn't really matter, then `Set` is ok.

Reply:
You're right, I fixed that